File: packages/omu/src/network/websocket-connection.ts
import { ByteReader, ByteWriter } from './bytebuffer.js';

import { Serializable } from '../serializer.js';
import type { Address } from './address.js';
import type { Connection } from './connection.js';
import { Packet, PacketData } from './packet/packet.js';

export class WebsocketConnection implements Connection {
    public connected: boolean;
    private socket: WebSocket | null;
    private readonly packetQueue: Array<PacketData> = [];
    private receiveWaiter: (() => void) | null = null;

    constructor(
        private readonly address: Address,
    ) {
        this.connected = false;
        this.socket = null;
    }

    private get wsEndpoint(): string {
        const protocol = this.address.secure ? 'wss' : 'ws';
        const { host, port } = this.address;
        return `${protocol}://${host}:${port}/ws`;
    }

    public connect(): Promise<void> {
        if (this.socket && !this.connected) {
            throw new Error('Already connecting');
        }
        return new Promise((resolve, reject) => {
            this.close();
            this.socket = new WebSocket(this.wsEndpoint);
            this.socket.onerror = reject;
            this.socket.onclose = () => this.close();
            this.socket.onmessage = (event) => this.onMessage(event);
            this.socket.onopen = () => {
                this.connected = true;
                resolve();
            }
        });
    }


    private async onMessage(event: MessageEvent<string | Blob>): Promise<void> {
        if (typeof event.data === 'string') {
            throw new Error('Received string data');
        }
        const reader = new ByteReader(await event.data.arrayBuffer());
        const type = reader.readString();
        const data = reader.readByteArray();
        reader.finish();
        this.packetQueue.push({ type, data });
        if (this.receiveWaiter) {
            this.receiveWaiter();
        }
    }

    public async receive(serializer: Serializable<Packet<unknown>, PacketData>): Promise<Packet> {
        if (this.receiveWaiter) {
            throw new Error('Already receiving');
        }
        while (this.packetQueue.length === 0) {
            await new Promise<void>((resolve) => {
                this.receiveWaiter = () => resolve();
            });
        }
        const packet = serializer.deserialize(this.packetQueue.shift()!);
        this.receiveWaiter = null;
        return packet;
    }

    get closed(): boolean {
        if (this.socket) {
            return this.socket.readyState === WebSocket.CLOSED;
        }
        return true;
    }

    close(): void {
        this.connected = false;
        if (this.socket) {
            this.socket.close();
            this.socket = null;
        }
    }

    send(packet: Packet, serializer: Serializable<Packet<unknown>, PacketData>): void {
        if (!this.connected || !this.socket) {
            throw new Error('Not connected');
        }
        const packetData = serializer.serialize(packet);
        const writer = new ByteWriter();
        writer.writeString(packetData.type);
        writer.writeByteArray(packetData.data);
        this.socket.send(writer.finish());
    }
}

File: packages-py/omu/src/omu/network/websocket_connection.py
from __future__ import annotations

from dataclasses import dataclass

import aiohttp
from aiohttp import ClientWebSocketResponse, web

from omu.address import Address
from omu.bytebuffer import ByteReader, ByteWriter
from omu.serializer import Serializable

from .connection import CloseError, Connection, Transport
from .packet import Packet, PacketData


@dataclass(frozen=True, slots=True)
class WebsocketsTransport(Transport):
    address: Address

    @property
    def _ws_endpoint(self) -> str:
        protocol = "wss" if self.address.secure else "ws"
        host = self.address.host or "127.0.0.1"
        port = self.address.port
        return f"{protocol}://{host}:{port}/ws"

    async def connect(self) -> WebsocketsConnection:
        session = aiohttp.ClientSession()
        socket = await session.ws_connect(self._ws_endpoint)
        return WebsocketsConnection(socket)


@dataclass(frozen=True, slots=True)
class WebsocketsConnection(Connection):
    socket: ClientWebSocketResponse

    async def send(self, packet: Packet, packet_mapper: Serializable[Packet, PacketData]) -> None:
        if not self.socket or self.socket.closed:
            raise RuntimeError("Not connected")
        packet_data = packet_mapper.serialize(packet)
        writer = ByteWriter()
        writer.write_string(packet_data.type)
        writer.write_uint8_array(packet_data.data)
        await self.socket.send_bytes(writer.finish())

    async def receive(self, packet_mapper: Serializable[Packet, PacketData]) -> Packet:
        if not self.socket or self.socket.closed:
            raise RuntimeError("Not connected")
        msg = await self.socket.receive()
        if msg.type in {
            web.WSMsgType.CLOSE,
            web.WSMsgType.CLOSED,
            web.WSMsgType.CLOSING,
            web.WSMsgType.ERROR,
        }:
            raise CloseError(f"Socket {msg.type.name.lower()}")
        if msg.data is None:
            raise RuntimeError("Received empty message")
        if msg.type == web.WSMsgType.TEXT:
            raise RuntimeError("Received text message")
        elif msg.type == web.WSMsgType.BINARY:
            with ByteReader(msg.data) as reader:
                event_type = reader.read_string()
                event_data = reader.read_uint8_array()
            packet_data = PacketData(event_type, event_data)
            return packet_mapper.deserialize(packet_data)
        else:
            raise RuntimeError(f"Unknown message type {msg.type}")

    async def close(self) -> None:
        if not self.socket or self.socket.closed:
            return
        await self.socket.close()

    @property
    def closed(self) -> bool:
        return not self.socket or self.socket.closed
